### ISSUE TITLE
no-useless-cast false positive

### ISSUE THREAD

**Issue by ethanresnick** [SonarSource/SonarTS#649]
**I want to report a bug.**

**SonarTS version:** 1.6.0
**Node.js version:** 8.x
**TypeScript version:** 2.8.3

**Rule key:** no-useless-cast

**Reproducer** 
```ts
const Identifier = (value: string) => ({ type: <"Identifier">"Identifier", value });
```

**Expected behavior**
The cast of `"Identifier"` to the literal string `"Identifier"` is marked as a useless cast, but it isn't: adding the cast changes the return type of the function to `{ type: "Identifier", value: string }` instead of `{ type: string, value: string }` because it stops Typescript from widening the type.

**Comment by andrea-guarino-sonarsource** [comment]
hi @ethanresnick ,
Thanks for raising the issue.
#650 has been created to handle this.
